distributed/rpc/rl/distributed_model.py

```diff
@@ -0,0 +1,287 @@
+import argparse
+import gym
+import numpy as np
+import os
+from itertools import chain, count
+
+import torch
+import torch.distributed.autograd as dist_autograd
+import torch.distributed.rpc as rpc
+import torch.multiprocessing as mp
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torch.distributed.rpc import RRef, rpc_sync, rpc_async, remote
+from torch.distributions import Categorical
+
+TOTAL_EPISODE_STEP = 5000
+AGENT_NAME = "agent"
+OBSERVER_NAME = "observer{}"
+
+parser = argparse.ArgumentParser(description='PyTorch RPC RL example')
+parser.add_argument('--world-size', type=int, default=2, metavar='W',
+                    help='world size for RPC, rank 0 is the agent, others are observers')
+parser.add_argument('--gamma', type=float, default=0.99, metavar='G',
+                    help='discount factor (default: 0.99)')
+parser.add_argument('--seed', type=int, default=543, metavar='N',
+                    help='random seed (default: 543)')
+parser.add_argument('--log-interval', type=int, default=10, metavar='N',
+                    help='interval between training status logs (default: 10)')
+args = parser.parse_args()
+
+torch.manual_seed(args.seed)
+
+
+def _call_method(method, rref, *args, **kwargs):
+    r"""
+    a helper function to call a method on the given RRef
+    """
+    return method(rref.local_value(), *args, **kwargs)
+
+
+def _remote_method(method, rref, *args, **kwargs):
+    r"""
+    a helper function to run method on the owner of rref and fetch back the
+    result using RPC
+    """
+    args = [method, rref] + list(args)
+    return rpc_sync(rref.owner(), _call_method, args=args, kwargs=kwargs)
+
+
+def _async_remote_method(method, rref, *args, **kwargs):
+    args = [method, rref] + list(args)
+    return rpc_async(rref.owner(), _call_method, args=args, kwargs=kwargs)
+
+class ObserverPolicy(nn.Module):
+    def __init__(self):
+        super(ObserverPolicy, self).__init__()
+        self.affine1 = nn.Linear(4, 128)
+        self.dropout = nn.Dropout(p=0.6)
+        self.affine2 = nn.Linear(128, 2)
+
+    def forward(self, x):
+        x = self.affine1(x)
+        x = self.dropout(x)
+        x = F.relu(x)
+        return self.affine2(x)
+
+
+class AgentPolicy(nn.Module):
+    def __init__(self):
+        super(AgentPolicy, self).__init__()
+
+        self.saved_log_probs = []
+        self.rewards = []
+
+    def forward(self, action_scores):
+        return F.softmax(action_scores, dim=1)
+
+
+class Observer:
+    def __init__(self):
+        self.id = rpc.get_worker_info().id
+        self.env = gym.make('CartPole-v1')
+        self.env.seed(args.seed)
+        self.policy = ObserverPolicy()
+
+    def get_gradients(self, ctx_id):
+        all_grads = dist_autograd.get_gradients(ctx_id)
```

Comment:
After the distributed backward pass, the grads for the `ObserverPolicy` model lives in `dist_autograd.get_gradients(ctx_id)` on the observer. Retrieve them and give them to the agent.